pep-9999.rst

```diff
@@ -0,0 +1,937 @@
+
+
+Illustration of the Deep Freezing Semantics
+-------------------------------------------
+
+Consider the following code:
+
+.. code-block:: python
+
+  class Foo:
+      pass
+  
+  x = Foo()
+  x.f = 42
+
+
+The ``Foo`` instance pointed to by ``x`` consists of several
+objects: its fields are stored in a dictionary object, and the
+assignment ``x.f = 42`` adds two objects to the dictionary in the
+form of a string key ``"f"`` and its associated value ``42``.
+These objects each have pointers to the ``string`` and ``int``
+type objects respectively. Similarly, the ``foo`` instance has a
+pointer to the ``Foo`` type object. Finally, all type objects have
+pointers to the same meta class object (``type``).
+
+Calling ``freeze(x)`` will freeze **all** of these objects.
+
+
+Default (Im)Mutabiliy 
+---------------------
+
+Except for the type object for ``NotFreezable``, no objects are
+immutable by default.
+
+**Rejected alternative**: Interned strings, numbers in the small
+integer cache, and tuples of immutable objects could be made
+immutable in this PEP. This is either consistent with current
+Python semantics or backwards-compatible. We have rejected this
+for now as we have not seen a strong need to do so. (A reasonable
+such design would make *all* numbers immutable, not just those in
+the small integer cache. This should be properly investigated.)
+
+
+Consequences of Deep Freezing
+=============================
+
+* The most obvious consequence of deep freezing is that it can lead
+  to surprising results when programmers fail to reason correctly
+  about the object structures in memory and how the objects reference
+  each other. For example, consider ``freeze(x)`` followed by
+  ``y.f = 42``. If the object in ``x`` can reach the same object that
+  ``y`` points to, then, the assignment will fail. **Mitigation:** To
+  facilitate debugging, exceptions due to attempting to mutate immutable
+  objects will include information about on what line an object was made
+  immutable.
+* Class Freezing: Freezing an instance of a user-defined class
+  will also freeze its class. Otherwise, sharing an immutable object
+  across threads would lead to sharing its *mutable* type object. Thus,
+  freezing an object also freezes the type type object of its super
+  classes. This means that any metaprogramming or changes to a class
+  must happen before a class is frozen. **Mitigation:** A frozen class
+  can be extended and its behaviour overridden through normal object-oriented
+  means. If neccessary, it is possible to add an option to make a mutable
+  copy of immutable objects and classes, which could then be changed.
+  Mutable instances of an immutable class can have their classes changed
+  to the mutable copy by reassigning ``__class__``.
+* Metaclass Freezing: Since class objects have metaclasses,
+  freezing a class may propagate upwards through the metaclass
+  hierarchy. This means that the ``Type`` object will be made immutable
```

Comment:
Above it was called `type` with a lower letter, which, I believe, is the correct way?

```suggestion
  hierarchy. This means that the ``type`` object will be made immutable
```